dash/include/dash/memory/internal/MemorySpaceRegistry.h

```diff
@@ -0,0 +1,60 @@
+#ifndef DASH__MEMORY__INTERNAL__POINTER_REGISTRY__INCLUDED_H
+#define DASH__MEMORY__INTERNAL__POINTER_REGISTRY__INCLUDED_H
+
+#include <memory>
+#include <mutex>
+#include <vector>
+
+#include <dash/Types.h>
+#include <dash/dart/if/dart_globmem.h>
+
+namespace dash {
+
+namespace internal {
+
+class MemorySpaceRegistry {
+  using segid_t  = int16_t;
+  using teamid_t = int16_t;
+
+  using key_t = std::pair<teamid_t, segid_t>;
+
+  static constexpr int IDX_TEAMID = 0;
+  static constexpr int IDX_SEGID  = 1;
+
+  using value_t = void*;
+
+public:
+  using element_t = std::pair<key_t, value_t>;
+
+  ~MemorySpaceRegistry() = default;
+  static MemorySpaceRegistry& GetInstance();
+
+  bool add(dart_gptr_t gptr, value_t mem_space);
+  void erase(dart_gptr_t gptr);
+
+  inline value_t lookup(dart_gptr_t pointer) const noexcept
+  {
+    auto it = do_lookup(std::make_pair(pointer.teamid, pointer.segid));
+
+    if (it != std::end(m_segments)) {
+      return it->second;
+    }
+
+    return nullptr;
+  }
+
+private:
+  static std::unique_ptr<MemorySpaceRegistry> m_instance;
+  static std::vector<element_t>               m_segments;
```

Comment:
Why not use a ` std::unordered_map` and hash `segid_t` and `teamid_t` into a 32-bit integer? The linear search may otherwise become significant for larger numbers of segments, e.g., when using the dynamic data structures.

Reply:
std::vector is way faster for a small number of entries than `unordered_map`. We can change it in future but for now I would stay with a `std::vector` because we have rarely more than 10 allocated global memory instances in a single DASH application.